libgraph/include/katana/TransformationView.h

```diff
@@ -0,0 +1,105 @@
+/// analytics routines as a regular property graph.
+class KATANA_EXPORT TransformationView final : public PropertyGraph {
+public:
+  TransformationView() = default;
+  TransformationView(TransformationView&& other) = default;
+
+  virtual ~TransformationView();
+
+  /// Make a projected graph from a property graph. Shares state with
+  /// the original graph.
+  static std::unique_ptr<TransformationView> MakeProjectedGraph(
+      const PropertyGraph& pg, const std::vector<std::string>& node_types,
+      const std::vector<std::string>& edge_types);
+
+  /// Return the number of nodes of the original property graph.
+  uint64_t NumOriginalNodes() const final {
+    return transformation_.original_to_transformed_nodes_.size();
+  }
+
+  /// Return the number of edges of the original property graph.
+  uint64_t NumOriginalEdges() const final {
+    return transformation_.original_to_transformed_edges_.size();
+  }
+
+  /// @param eid the input eid (must be projected edge id)
+  Edge TransformedToOriginalEdgeId(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < NumEdges());
+    return topology().GetLocalEdgeIDFromOutEdge(eid);
+  }
+
+  /// @param nid the input node id (must be projected node id)
+  Node TransformedToOriginalNodeId(const Node& nid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(nid < NumNodes());
+    return topology().GetLocalNodeID(nid);
+  }
+
+  /// @param eid the input eid (must be original edge id)
+  Edge OriginalToTransformedEdgeId(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < NumEdges());
+    return transformation_.original_to_transformed_edges_[eid];
+  }
+
+  /// @param nid the input node id (must be original node id)
+  Node OriginalToTransformedNodeId(const Node& nid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(nid < NumNodes());
+    return transformation_.original_to_transformed_nodes_[nid];
+  }
+
+  /// Bitmask of nodes included in the transformation view. Should be used to construct arrow tables.
+  std::shared_ptr<arrow::Buffer> NodeBitmask() const noexcept {
+    auto length = transformation_.original_to_transformed_nodes_.size();
+    return std::make_shared<arrow::Buffer>(
+        transformation_.node_bitmask_data_.data(),
+        arrow::BitUtil::BytesForBits(length));
+  }
+
+  /// Bitmask of edges included in the transformation view. Should be used to construct arrow tables.
+  std::shared_ptr<arrow::Buffer> EdgeBitmask() const noexcept {
+    auto length = transformation_.original_to_transformed_edges_.size();
+    return std::make_shared<arrow::Buffer>(
+        transformation_.edge_bitmask_data_.data(),
+        arrow::BitUtil::BytesForBits(length));
+  }
+
+private:
+  struct Transformation {
+    NUMAArray<Node> original_to_transformed_nodes_;
+    NUMAArray<Edge> original_to_transformed_edges_;
+    NUMAArray<uint8_t> node_bitmask_data_;
+    NUMAArray<uint8_t> edge_bitmask_data_;
```

Comment:
Bitmask data needs to be in PropertyGraph, otherwise we won't be able to do Transformations of Projections properly.

E.g. 
```
PropertyGraph proj_pg = TranformationView::MakeProjection(orig_pg);
/// will be a transpose on the full graph rather than the projection
PropertyGraph tansposed_pg = TransformationView::MakeTransposed(proj_pg);
```

Reply:
To follow up on this, I think a naïve implementation of a bitmask of 1s for the un-transformed PG is probably fine for now, and we can optimize it out with either a compressed bitmap (preferred) or a std::optional<bitmap> in the future

Reply:
I don't follow, I'm afraid. I'll ask you on Monday in person :-)